### ISSUE TITLE
Layer 0 is not independent from layer 1 with respect to 320x480 and mode 7 (8bpp)

### ISSUE THREAD

**Issue by MCPEmaniac** [commanderx16/x16-emulator#190]
When layer 0 is set for 320x480 and mode 7 (8bpp) it doesn't work properly unless layer 1 is also set to 320x480. When filling the layer 0 screen with a color, only about half the layer 0 screen is filled unless layer 1 is also set to 320x480 even with layer 1 disabled.

**Comment by jjbliss** [comment]
I haven't tested this, but here is my understanding of how it should work.

Let's say you set the composer to run at 640x480 mode.
You can load a 640x480 bitmap into layer 0.
load a 320x240 bitmap into layer1.
Use the HSCROLL and VSCROLL on layer 1 to move layer1 bitmap across layer 0.

Or, you could have a 640x480 bitmap when the composer is in 320x240 mode.
Then using HSCROLL and VSCROLL you can pan across the image.

**Comment by indigodarkwolf** [comment]
I'm not sure there's an issue here. The resolution multiplier is set by the composer through the HSCALE and VSCALE parameters, of course this impacts both layers, and all 128 sprites as well.

There isn't even enough memory for 320x480x8bpp anyways. You run out after line 409.

In practice, I don't expect you would worry about more than 1 layer in bitmap mode. I mean, you could set the composer for 320x240 resolution, and would then be able to have two layers at 320x240x4bpp, but why? You couldn't double-buffer them, and you'd get a much larger color space working with layer 0 at 320x240x8bpp. And you don't need two layers just to double-buffer, in fact I suspect that's doing it the way that is both slower and harder than adjusting the tile data offset.

At that point, think of layer 1 as a "bonus" that could toss in some tiled graphics with the unused 54,272 bytes, which is ample for, say, a 64x32 tileset (4KiB) of up to 784 distinct 8x8 tiles. Or trade some of that for sprites.

Maybe you could use layer 1 as a bitmap to support some kind of very fancy stencil over a tile-mode layer 0? But unless the visible areas underneath are less than a 64x64 pixel area, I expect it would still be more efficient to represent the stencil layer as a tilemap with distinct tiles.

**Comment by mist64** [comment]
Yes, HSCALE and VSCALE are a global property. If you set a scale of 1, all numbers of layers 0 and 1 are as documented. If you set a scale of two, you need to multiply all numbers in the layer 0/1 docs with 2.

Close?

**Comment by marksassistants0** [comment]
I agree




On Monday, October 28, 2019, 8:14 PM, Stephen Horn <notifications@github.com> wrote:


I'm not sure there's an issue here. The resolution multiplier is set by the composer through the HSCALE and VSCALE parameters, of course this impacts both layers, and all 128 sprites as well.

There isn't even enough memory for 320x480x8bpp anyways. You run out after line 409.

In practice, I don't expect you would worry about more than 1 layer in bitmap mode. I mean, you could set the composer for 320x240 resolution, and would then be able to have two layers at 320x240x4bpp, but why? You couldn't double-buffer them, and you'd get a much larger color space working with layer 0 at 320x240x8bpp. And you don't need two layers just to double-buffer, in fact I suspect that's doing it the way that is both slower and harder than adjusting the tile data offset.

At that point, think of layer 1 as a "bonus" that could toss in some tiled graphics with the unused 54,272 bytes, which is ample for, say, a 64x32 tileset (4KiB) of up to 784 distinct 8x8 tiles. Or trade some of that for sprites.

Maybe you could use layer 1 as a bitmap to support some kind of very fancy stencil over a tile-mode layer 0? But unless the visible areas underneath are less than a 64x64 pixel area, I expect it would still be more efficient to represent the stencil layer as a tilemap with distinct tiles.

**Comment by oziphantom** [comment]
The Bitmap width setting is how many pixels it takes to be on a line. So when you set it to 640 wide memory location BASE+640 is the start of the next line. It sets the Stride not the visible width. What it would do with a 320 wide bitmap on a 640 wide display is a matter of working out how the VERA works it out in hardware, it would either start from the line again for the 2nd half of the screen or it would draw the "2nd" line on the right half of the screen then draw it again at the start of the 2nd line. Hard to tell if this is a bug until it can be compared to final hardware.

**Comment by ekmett** [comment]
> In practice, I don't expect you would worry about more than 1 layer in bitmap mode.

It is precisely because the 320x240 mode just barely fits that you have to get creative.

I'm currently playing with a toy project that is based almost entirely on exploiting 2 layers of bitmap mode graphics.

Layer 0 provides fairly static backdrops. (Eric Chahi's Another World used the ability to store 2 backgrounds off to the side for this to save redraw.) Overlays double buffer, so I don't have to torture myself to make the target frame rate (which is good, because it is mostly polygons, and I'll be lucky to get 20fps.) Currently using 4bpp for all 3 layers, but if I crop the vertical to 200 or find a way to squeeze my content into 2bpp for the fg layers by, say, using more sprites for rarer colors, I could go to 8bpp for the backdrop layer. (Using 4bpp ~doubles the speed of polygon blitting, and moving to 2bpp would do so again.) That way I can save all the effort blitting the background over and over, and the foreground layers just superimpose, and I can clear them at my leisure and flip whenever a vblank comes around.

Most of my joy in this platform has come from playing with the weird combinations possible with the VERA.

You can also play some weird tricks where you use 3 separate 320x240 8bpp displays that overlap, so that you can double buffer _part_ of the screen, e.g. you have enough video ram for 320x409 + a few toy sprites, so you could set a foreground layer up at base 320*168, and ping pong layer 0 between base 0 and base 320*84 and so long as everything in the bottom half of the screen on layer 1 was opaque you can use that to double buffer the top 84 lines f the screen. Pretty awful, but we all played games that had huge cockpits and the like to reduce fill rate requirements back in the day.
